### ISSUE TITLE
K8s: Raft election wrong IP at every restart

### ISSUE THREAD

**Issue by Montbuet** [hashicorp/consul-k8s#561]
Hello,

I have a consul cluster deployed in a kubernetes cluster.
When stopping the Consul cluster and starting it again, every pod has a new IP address, and the raft election fails:

[ERROR] agent.server.raft: failed to make requestVote RPC: target="{Voter b91......... 172.XX.XX.XX.:8300}" error="dial tcp <nil>->72.XX.XX.XX.:8300: i/o timeout"
(172.XX.XX.XX is assign to no one)

The only thing I can do is to delete every volume and restore a snapshot file, and everything is good again after that. But it is obviously time consuming.

How can I tell consul that the IP address of each pod changes at every restart ? Is there anything I can write in my values helm chart to do that ?

Ty and have a nice day

**Comment by Montbuet** [comment]
Hi @david-yu ,

**Helm config file:**
`consul:
  global:
    enabled: true
    name: "consul"
    datacenter: "K8S"
    tls:
      enabled: false
      httpsOnly: false

  client:
    enabled: true
    securityContext:
      runAsNonRoot: false
      runAsGroup: 0
      runAsUser: 0
      fsGroup: 0

  server:
    replicas: 3
    bootstrap_expect: 3
    enabled: true
    securityContext:
      runAsNonRoot: false
      runAsGroup: 0
      runAsUser: 0
      fsGroup: 0
  ui:
    enabled: true`

**Reproduce steps:**
  * Consuls servers and agents are deployed successfully on the cluster, composed of 9 workers.
  * Helm -n vault delete vault-test (Uninstalling both Vault and Consul).
  * Redeploy it again on the cluster, the consuls pods are running on random k8s workers (not the same as previously).
  * Logs --> Consul agent election errors, trying to contact the old IP from the first deployment.

**What to do to:**
  - Tell Consul to rather use the dns name than the IPs (like consul-server-0, consul-server-1 and consul-server-2) ?
  - Tell Consul to be smarter and that the IP addresses are not static, and it need to perform a "discovery" in the current namespace ?

If you have any clue on how to do that, I would be very grateful.

**Comment by ishustava** [comment]
Hey @Montbuet 

What is the use case for that? Are you thinking of a disaster case when all pods go down? Right now, Consul (or rather Raft) cannot handle IP address change when the cluster does not have a quorum.

It looks like there was already some [discussion](https://github.com/hashicorp/consul/issues/1580#issuecomment-319402908) on changing raft to support DNS addresses or in general accommodate the use case when the entire cluster goes down, but I don't think this was addressed yet.

@david-yu If the request is to make Consul support IP address change for the entire cluster, then I think this issue is more appropriate in hashicorp/consul as consul-helm behaves as expected.

**Comment by Montbuet** [comment]
> Hey @Montbuet
> 
> What is the use case for that? Are you thinking of a disaster case when all pods go down? Right now, Consul (or rather Raft) cannot handle IP address change when the cluster does not have a quorum.
> 
> It looks like there was already some [discussion](https://github.com/hashicorp/consul/issues/1580#issuecomment-319402908) on changing raft to support DNS addresses or in general accommodate the use case when the entire cluster goes down, but I don't think this was addressed yet.
> 
> @david-yu If the request is to make Consul support IP address change for the entire cluster, then I think this issue is more appropriate in hashicorp/consul as consul-helm behaves as expected.

Hey and thank you for your answer.
Yes, this is precisely for a disaster recovery use case. Doing it by hand is time consuming, and in case something bad happen, it would be great that everything starts up without human intervention,

**Comment by ishustava** [comment]
Makes sense. I'll transfer it back to Consul so that it can be better tracked there. Sorry for all the back and forth!

For now, I think your best bet is to work around this manually or with scripting as this is likely a more involved change.
